### PR TITLE
Fix build issues when creating deb packages

### PR DESCRIPTION
**Pull request by hkhonming**
#### Problem solved by the commit
Fixed build failures for build_deb_edge.sh

#### How problem was solved, alternative solutions (if any) and why they were rejected
1. debian packaging skills is outdated and I just synchronise it with the latest packaging script
2. Fix the bug in dkms-edge.cmake introduced after the removal of cu_plram.c
3. Update to build on jammy by default 

#### Risks (if any) associated the changes in the commit
Most changes are specific to debian packaging and so it should be safe for other usage. The dkms-edge.cmake is an obvious bug after a removal of the cu_plram.c and so the fix is pretty straightforward too.

#### What has been tested and how, request additional testing if necessary
Tested to generate .deb file in 2022.1 branch head. If we need to cherrypick the change to master branch. Probably we need to test it too since I am not sure if there are other changes in master branch that require packaging script change.

#### Documentation impact (if any)
If there are document on how to create .deb files. We should update it and remove the need of the sysroot, as well as mentioning we are building jammy by default now.

### CODE REVIEW

**Comment by chvamshi-xilinx** [comment]
ok to test

**Review by stsoe — CHANGES_REQUESTED** [review]
I am confused.  2022.1 did not support jammy.  
Is this meant for master?

File: build/build_edge_deb.sh [hunk 1]
```diff
@@ -128,13 +123,13 @@ cp -rf $DEBIAN $DEBIAN_ARTIFACTS
...
-mv xrt-zocl-dkms_${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_arm64.deb xrt_zocl_dkms_${RELEASE_VERSION}.${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_20.04-arm64.deb
-mv xrt-embedded-dbgsym_${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_arm64.ddeb xrt_embedded_dbgsym_${RELEASE_VERSION}.${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_20.04-arm64.ddeb
+mv xrt-embedded_${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_arm64.deb xrt_embedded_${RELEASE_VERSION}.${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_22.04-arm64.deb
+mv xrt-zocl-dkms_${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_arm64.deb xrt_zocl_dkms_${RELEASE_VERSION}.${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_22.04-arm64.deb
+mv xrt-embedded-dbgsym_${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_arm64.ddeb xrt_embedded_dbgsym_${RELEASE_VERSION}.${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_22.04-arm64.ddeb
```
Comment (keryell) [inline build/build_edge_deb.sh:132]: I have the feeling we should put these `22.04` and `jammy` in some variable so it is easier to change or even configure (`22.10 is coming soon ;-) ).

**Comment by dayeh-xilinx** [comment]
retest this please.

File: build/build_edge_deb.sh (hunk 1 above)
Comment (hkhonming) [reply build/build_edge_deb.sh]: I have pushed a commit that add the DIST argument

**Comment by hkhonming** [comment]
@stsoe During the enablement in jammy, the latest version of XRT packaged is 2.13.466. As far as I know it should be coming from the 2022.1 branch

File: build/build_edge_deb.sh [hunk 2]
```diff
@@ -128,13 +127,13 @@ cp -rf $DEBIAN $DEBIAN_ARTIFACTS
...
-mv xrt-zocl-dkms_${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_arm64.deb xrt_zocl_dkms_${RELEASE_VERSION}.${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_20.04-arm64.deb
-mv xrt-embedded-dbgsym_${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_arm64.ddeb xrt_embedded_dbgsym_${RELEASE_VERSION}.${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_20.04-arm64.ddeb
+mv xrt-embedded_${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_arm64.deb xrt_embedded_${RELEASE_VERSION}.${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_${DIST}-arm64.deb
+mv xrt-zocl-dkms_${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_arm64.deb xrt_zocl_dkms_${RELEASE_VERSION}.${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_${DIST}-arm64.deb
+mv xrt-embedded-dbgsym_${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_arm64.ddeb xrt_embedded_dbgsym_${RELEASE_VERSION}.${XRT_MAJOR_VERSION}.${XRT_MINOR_VERSION}.${XRT_VERSION_PATCH}_${DIST}-arm64.ddeb
```
Comment (keryell) [inline build/build_edge_deb.sh:136]: So now the packages have a `_jammy-arm64.deb` extension instead of `_22.04-arm64.deb`?

Comment (hkhonming) [reply build/build_edge_deb.sh:136]: Yes. I thought it is more straightforward. The originally code hardcode the argument to sbuild to "focal" and do the rename with suffix "20.04". So I guess it is more straightforward to just name it by the distribution you passed.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Comment by dayeh-xilinx** [comment]
retest this please.

**Review by chvamshi-xilinx — APPROVED** [review]

**Comment by rbramand-xilinx** [comment]
Hi @hkhonming, one query did you use cross compilation for testing this? (host - x64 or amd64)
because when I tested on amd64 host it fails if I don't provide sysroot

**Comment by hkhonming** [comment]
@rbramand-xilinx The original script depends on the setup of sbuild chroot & the xilinx sysroot. I am removing the sysroot since XRT itself actually just depends on standard libraries. In this case, you will still need to setup a standard sbuild chroot in arm64 architecture.

**Comment by rbramand-xilinx** [comment]
> @rbramand-xilinx The original script depends on the setup of sbuild chroot & the xilinx sysroot. I am removing the sysroot since XRT itself actually just depends on standard libraries. In this case, you will still need to setup a standard sbuild chroot in arm64 architecture.

@hkhonming I am confused, you are saying that sbuild chroot in arm64 is needed but removed sysroot option, how does that work? Also we would like to do cross compilation mostly, so we need to provide sysroot flag right?

**Comment by hkhonming** [comment]
> > @rbramand-xilinx The original script depends on the setup of sbuild chroot & the xilinx sysroot. I am removing the sysroot since XRT itself actually just depends on standard libraries. In this case, you will still need to setup a standard sbuild chroot in arm64 architecture.
> 
> @hkhonming I am confused, you are saying that sbuild chroot in arm64 is needed but removed sysroot option, how does that work? Also we would like to do cross compilation mostly, so we need to provide sysroot flag right?

The original setup used sbuild-createchroot to pull the sysroot to the chroot environment. Since there are no specific dependencies on Xilinx packages. You can just use normal mk-sbuild jammy/focal --arch=arm64 to create the chroot. Then sbuild should be able to build XRT through qemu.

Ref: https://packaging.ubuntu.com/html/setting-up-sbuild.html

File: build/build_edge_deb.sh (hunk 2 above)
Comment (rozumx) [reply build/build_edge_deb.sh:137]: Hmm...Using OS release names instead of versions doesn't sound like very scalable.

File: build/debian/rules [hunk 3]
```diff
@@ -16,4 +16,21 @@ override_dh_auto_test:
 	dh_auto_test || true
 
 override_dh_shlibdeps:
-	dh_shlibdeps -l/opt/xilinx/xrt/lib
+	dh_shlibdeps -l/usr/lib -l/usr/lib/xrt/module
+
+override_dh_install:
+# TODO Some files are missing in tmp, this will need a deeper investigation
```
Comment (rozumx) [inline build/debian/rules:22]: What files are missing????  And how does this impact the installation?

**Review by rozumx — CHANGES_REQUESTED** [review]

**Comment by keryell** [comment]
> I am confused. 2022.1 did not support jammy. Is this meant for master?

I was just about asking the same.
Is there someone expected to also add some documentation about how to use this?

**Comment by keryell** [comment]
> Hi, I will send a fresh PR for master and address comments on this PR. This PR is mainly for **Edge** platforms. Regarding documentation, at present we are using this internally and I will create a confluence page with steps on creating packages using the script. We can add the steps to XRT docs based on marketing ask

This is open-source code and we need also some documentation on how to build this on Edge platforms. This is what we are looking at right now on our ZCU102.
Why a new PR? Just go on working on it.
Globally open-sourcing tends to increase the software quality.
@Ralender: could you add your `perl` issues here?

**Comment by keryell** [comment]
@Ralender: ping.

**Comment by chvamshi-xilinx** [comment]
Hi @hkhonming , You can close this PR as https://github.com/Xilinx/XRT/pull/7148 is merged to master XRT,
Users can use master XRT branch to build debian packages.

**Comment by hkhonming** [comment]
Supersede by https://github.com/Xilinx/XRT/pull/7148

**Pull request closed without merging**